interfaces/prompting/patterns/variant.go

```diff
@@ -0,0 +1,428 @@
+	compSeparatorDoublestar
+	compTerminal // marker of end of pattern.
+	compSeparator
+	compAnySingle // ? has precedence over / so that /foo*?/ has precedence over /foo*/
+	compLiteral
+)
+
+type component struct {
+	compType componentType
+	compText string
+}
+
+// String returns the globstar-style pattern string associated with the given
+// component.
+func (c component) String() string {
+	switch c.compType {
+	case compGlobstar:
+		return "*"
+	case compSeparatorDoublestarTerminal:
+		return "/**"
+	case compSeparatorDoublestarSeparatorTerminal:
+		return "/**/"
+	case compSeparatorDoublestar:
+		return "/**"
+	case compTerminal: // end of pattern
+		return ""
+	case compSeparator:
+		return "/"
+	case compAnySingle:
+		return "?"
+	case compLiteral:
+		return c.compText
+	}
+	return "###ERR_UNKNOWN_COMPONENT_TYPE###" // Should not occur
+}
+
+// componentRegex returns a regular expression corresponding to the bash-style
+// globstar matching behavior of the receiving component.
+//
+// For example, "*" matches any non-separator characters, so we return the regex
+// `((?:[^/]|\\/)*)` for the globstar component type.
+//
+// The returned regexps should each be enclosed in a capturing group with no
+// capturing groups within. This allows a single regex to be constructed for a
+// given pattern variant by concatenating all the component regular expressions
+// together, and the resulting regex has exactly one capturing group for each
+// component, in order.
+func (c component) componentRegex() string {
+	switch c.compType {
+	case compGlobstar:
+		return `((?:[^/]|\\/)*)`
+	case compSeparatorDoublestarTerminal:
+		return `((?:/.+)?/?)`
+	case compSeparatorDoublestarSeparatorTerminal:
+		return `((?:/.+)?/)`
+	case compSeparatorDoublestar:
+		return `((?:/.+)?)`
+	case compTerminal:
+		return `(/?)`
+	case compSeparator:
+		return `(/)`
+	case compAnySingle:
+		return `([^/])` // does escaped '/' (e.g. `\\/`) count as one character?
+	case compLiteral:
+		return `(` + regexp.QuoteMeta(unescapeLiteral(c.compText)) + `)`
+	}
+	return `()`
+}
+
+var escapeFinder = regexp.MustCompile(`\\(.)`)
+
+// unescapeLiteral removes any `\` characters which are used to escape another
+// character. Note that escaped `\` characters are not removed, since they are
+// not acting as an escape character in those instances. That is, `\\` is
+// reduced to `\`.
+func unescapeLiteral(literal string) string {
+	return escapeFinder.ReplaceAllString(literal, "${1}")
+}
+
+type PatternVariant struct {
+	variant    string
+	components []component
+	regex      *regexp.Regexp
+}
+
+// String returns the rendered string associated with the pattern variant.
+func (v PatternVariant) String() string {
+	return v.variant
+}
+
+// ParsePatternVariant parses a rendered variant string into a PatternVariant
+// whose precedence can be compared against others.
+func ParsePatternVariant(variant string) (PatternVariant, error) {
```

Comment:
it's a bit unclear to me why this is exported?

Reply:
I agree, I don't think it needs to be.